src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -97,169 +130,363 @@ func newCommitLogSource(opts Options, inspection fs.Inspection) bootstrap.Source
 // TODO(rartoul): Make this take the SnapshotMetadata files into account to reduce the
 // number of commitlogs / snapshots that we need to read.
-func (s *commitLogSource) ReadData(
-	ns namespace.Metadata,
-	shardsTimeRanges result.ShardTimeRanges,
-	runOpts bootstrap.RunOptions,
-) (result.DataBootstrapResult, error) {
-	if shardsTimeRanges.IsEmpty() {
-		return result.NewDataBootstrapResult(), nil
+func (s *commitLogSource) Read(
+	namespaces bootstrap.Namespaces,
+) (bootstrap.NamespaceResults, error) {
+	timeRangesEmpty := true
+	for _, elem := range namespaces.Namespaces.Iter() {
+		namespace := elem.Value()
+		dataRangesNotEmpty := !namespace.DataRunOptions.ShardTimeRanges.IsEmpty()
+
+		indexEnabled := namespace.Metadata.Options().IndexOptions().Enabled()
+		indexRangesNotEmpty := indexEnabled && !namespace.IndexRunOptions.ShardTimeRanges.IsEmpty()
+		if dataRangesNotEmpty || indexRangesNotEmpty {
+			timeRangesEmpty = false
+			break
+		}
+	}
+	if timeRangesEmpty {
+		// Return empty result with no unfulfilled ranges.
+		return bootstrap.NewNamespaceResults(namespaces), nil
 	}
 
 	var (
 		// Emit bootstrapping gauge for duration of ReadData
-		doneReadingData        = s.metrics.data.emitBootstrapping()
-		encounteredCorruptData = false
-		fsOpts                 = s.opts.CommitLogOptions().FilesystemOptions()
-		filePathPrefix         = fsOpts.FilePathPrefix()
+		doneReadingData         = s.metrics.emitBootstrapping()
+		encounteredCorruptData  = false
+		fsOpts                  = s.opts.CommitLogOptions().FilesystemOptions()
+		filePathPrefix          = fsOpts.FilePathPrefix()
+		namespaceResults        = make(map[string]*readNamespaceResult)
+		setInitialTopologyState bool
+		initialTopologyState    *topology.StateSnapshot
 	)
 	defer doneReadingData()
 
-	// Determine which snapshot files are available.
-	snapshotFilesByShard, err := s.snapshotFilesByShard(
-		ns.ID(), filePathPrefix, shardsTimeRanges)
-	if err != nil {
-		return nil, err
+	startSnapshotsRead := s.nowFn()
+	s.log.Info("read snapshots start")
+	for _, elem := range namespaces.Namespaces.Iter() {
+		ns := elem.Value()
+		accumulator := ns.DataAccumulator
+
+		// NB(r): Combine all shard time ranges across data and index
+		// so we can do in one go.
+		shardTimeRanges := result.ShardTimeRanges{}
+		shardTimeRanges.AddRanges(ns.DataRunOptions.ShardTimeRanges)
+		if ns.Metadata.Options().IndexOptions().Enabled() {
+			shardTimeRanges.AddRanges(ns.IndexRunOptions.ShardTimeRanges)
+		}
+
+		namespaceResults[ns.Metadata.ID().String()] = &readNamespaceResult{
+			namespace:               ns,
+			dataAndIndexShardRanges: shardTimeRanges,
+		}
+
+		// Make the initial topology state available.
+		if !setInitialTopologyState {
+			setInitialTopologyState = true
+			initialTopologyState = ns.DataRunOptions.RunOptions.InitialTopologyState()
+		}
+
+		// Determine which snapshot files are available.
+		snapshotFilesByShard, err := s.snapshotFilesByShard(
```

Comment:
May be better as a TODO, but would it be worth getting the files for each namespace and shard first to avoid scanning disk per ns per shard?